File: zig/src/udp_listener.zig
//! UDP Listener Transport implementation.
//!
//! Provides an unconnected UDP transport that can send to and receive from
//! multiple remote addresses. Suitable for Host which manages multiple peers.

const std = @import("std");
const posix = std.posix;
const net = std.net;
const transport = @import("transport.zig");
const udp = @import("udp.zig");

/// UDP Listener over an unconnected UDP socket.
///
/// Can send to and receive from multiple remote addresses.
/// Suitable for Host which manages multiple peers on a single port.
pub const UdpListener = struct {
    fd: posix.fd_t,
    local_addr: udp.UdpAddr,
    closed: std.atomic.Value(bool),

    const Self = @This();

    /// Create a new UDP listener bound to the specified address.
    ///
    /// Use "0.0.0.0:0" to let the OS assign an available port.
    /// Use "0.0.0.0:51820" to bind to a specific port.
    pub fn bind(bind_addr_str: []const u8) !Self {
        const bind_addr = try udp.UdpAddr.parse(bind_addr_str);

        // Create socket
        const fd = try posix.socket(posix.AF.INET, posix.SOCK.DGRAM, 0);
        errdefer posix.close(fd);

        // Bind
        try posix.bind(fd, &bind_addr.inner.any, bind_addr.inner.getOsSockLen());

        // Get actual bound address
        var bound_addr: net.Address = undefined;
        var bound_len: posix.socklen_t = @sizeOf(posix.sockaddr);
        try posix.getsockname(fd, &bound_addr.any, &bound_len);

        return .{
            .fd = fd,
            .local_addr = udp.UdpAddr.init(bound_addr),
            .closed = std.atomic.Value(bool).init(false),
        };
    }

    /// Get the local port.
    pub fn port(self: Self) u16 {
        return self.local_addr.port();
    }

    /// Get the local address.
    pub fn localAddrUdp(self: Self) udp.UdpAddr {
        return self.local_addr;
    }

    /// Check if closed.
    pub fn isClosed(self: *Self) bool {
        return self.closed.load(.seq_cst);
    }

    /// Send data to the specified address.
    pub fn sendTo(self: *Self, data: []const u8, to_addr: transport.Addr) transport.TransportError!void {
        if (self.closed.load(.seq_cst)) {
            return error.Closed;
        }

        const dest = switch (to_addr) {
            .udp => |a| a.inner,
            .mock => return error.InvalidAddress,
        };

        _ = posix.sendto(self.fd, data, 0, &dest.any, dest.getOsSockLen()) catch {
            return error.IoError;
        };
    }

    /// Receive data and return the sender's address.
    pub fn recvFrom(self: *Self, buf: []u8) transport.TransportError!transport.RecvResult {
        if (self.closed.load(.seq_cst)) {
            return error.Closed;
        }

        var from_addr: posix.sockaddr = undefined;
        var from_len: posix.socklen_t = @sizeOf(posix.sockaddr);

        const n = posix.recvfrom(self.fd, buf, 0, &from_addr, &from_len) catch |err| {
            if (err == error.WouldBlock) {
                return error.IoError; // Timeout
            }
            return error.IoError;
        };

        // Convert sockaddr to net.Address
        const addr = net.Address{ .any = from_addr };

        return .{
            .bytes_read = n,
            .from_addr = (udp.UdpAddr.init(addr)).toAddr(),
        };
    }

    /// Get local address via Transport interface.
    pub fn localAddr(self: *Self) transport.Addr {
        return self.local_addr.toAddr();
    }

    /// Close the transport.
    pub fn close(self: *Self) void {
        if (!self.closed.swap(true, .seq_cst)) {
            posix.close(self.fd);
        }
    }

    /// Set receive timeout in milliseconds.
    pub fn setRecvTimeout(self: Self, timeout_ms: u32) !void {
        const tv = posix.timeval{
            .sec = @intCast(timeout_ms / 1000),
            .usec = @intCast((timeout_ms % 1000) * 1000),
        };
        try posix.setsockopt(self.fd, posix.SOL.SOCKET, posix.SO.RCVTIMEO, std.mem.asBytes(&tv));
    }

    /// Get Transport vtable.
    pub fn asTransport(self: *Self) transport.Transport {
        return .{
            .ptr = self,
            .vtable = &.{
                .send = @ptrCast(&sendToTransport),
                .recv = @ptrCast(&recvFromTransport),
                .local_addr = @ptrCast(&localAddrTransport),
                .close = @ptrCast(&closeTransport),
            },
        };
    }

    // Transport vtable implementations
    fn sendToTransport(self: *Self, data: []const u8, addr: transport.Addr) transport.TransportError!void {
        return self.sendTo(data, addr);
    }

    fn recvFromTransport(self: *Self, buf: []u8) transport.TransportError!transport.RecvResult {
        return self.recvFrom(buf);
    }

    fn localAddrTransport(self: *Self) transport.Addr {
        return self.localAddr();
    }

    fn closeTransport(self: *Self) transport.TransportError!void {
        self.close();
    }
};

// =============================================================================
// Tests
// =============================================================================

test "UdpListener.bind" {
    var listener = try UdpListener.bind("127.0.0.1:0");
    defer listener.close();

    try std.testing.expect(listener.port() > 0);
    try std.testing.expect(!listener.isClosed());
}

test "UdpListener send/recv" {
    var listener1 = try UdpListener.bind("127.0.0.1:0");
    defer listener1.close();

    var listener2 = try UdpListener.bind("127.0.0.1:0");
    defer listener2.close();

    const addr2 = listener2.localAddr();

    // Send from listener1 to listener2
    const data = "hello world";
    try listener1.sendTo(data, addr2);

    // Receive on listener2
    var buf: [1024]u8 = undefined;
    const result = try listener2.recvFrom(&buf);

    try std.testing.expectEqualStrings(data, buf[0..result.bytes_read]);
}

test "UdpListener multiple peers" {
    var server = try UdpListener.bind("127.0.0.1:0");
    defer server.close();

    var client1 = try UdpListener.bind("127.0.0.1:0");
    defer client1.close();

    var client2 = try UdpListener.bind("127.0.0.1:0");
    defer client2.close();

    const server_addr = server.localAddr();

    // Both clients send to server
    try client1.sendTo("from client1", server_addr);
    try client2.sendTo("from client2", server_addr);

    // Server receives both
    var buf: [1024]u8 = undefined;
    var msg1_buf: [32]u8 = undefined;
    var msg2_buf: [32]u8 = undefined;
    var msg1: []u8 = &.{};
    var msg2: []u8 = &.{};

    const result1 = try server.recvFrom(&buf);
    @memcpy(msg1_buf[0..result1.bytes_read], buf[0..result1.bytes_read]);
    msg1 = msg1_buf[0..result1.bytes_read];

    const result2 = try server.recvFrom(&buf);
    @memcpy(msg2_buf[0..result2.bytes_read], buf[0..result2.bytes_read]);
    msg2 = msg2_buf[0..result2.bytes_read];

    // Check both messages received (order may vary)
    const found1 = std.mem.eql(u8, msg1, "from client1") or std.mem.eql(u8, msg2, "from client1");
    const found2 = std.mem.eql(u8, msg1, "from client2") or std.mem.eql(u8, msg2, "from client2");

    try std.testing.expect(found1);
    try std.testing.expect(found2);
}

File: rust/src/udp_listener.rs
//! UDP Listener Transport implementation.
//!
//! Provides an unconnected UDP transport that can send to and receive from
//! multiple remote addresses. Suitable for Host which manages multiple peers.

use std::io;
use std::net::{SocketAddr, ToSocketAddrs, UdpSocket};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Mutex;

use crate::transport::{Addr, Result, Transport, TransportError};
use crate::udp::UdpAddr;

/// UDP Listener over an unconnected UDP socket.
///
/// Can send to and receive from multiple remote addresses.
/// Suitable for Host which manages multiple peers on a single port.
pub struct UdpListener {
    socket: UdpSocket,
    local_addr: UdpAddr,
    closed: AtomicBool,
    // Mutex for thread-safe close
    close_lock: Mutex<()>,
}

impl UdpListener {
    /// Create a new UDP listener bound to the specified address.
    ///
    /// Use "0.0.0.0:0" to let the OS assign an available port.
    /// Use "0.0.0.0:51820" to bind to a specific port.
    pub fn bind(bind_addr: &str) -> io::Result<Self> {
        let addr = bind_addr
            .to_socket_addrs()?
            .next()
            .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, "no address found"))?;

        let socket = UdpSocket::bind(addr)?;
        let local_addr = UdpAddr::new(socket.local_addr()?);

        Ok(Self {
            socket,
            local_addr,
            closed: AtomicBool::new(false),
            close_lock: Mutex::new(()),
        })
    }

    /// Get the local address.
    pub fn local_addr_udp(&self) -> &UdpAddr {
        &self.local_addr
    }

    /// Get the local port.
    pub fn port(&self) -> u16 {
        self.local_addr.socket_addr().port()
    }

    /// Set read timeout.
    pub fn set_read_timeout(&self, timeout: Option<std::time::Duration>) -> io::Result<()> {
        self.socket.set_read_timeout(timeout)
    }

    /// Set write timeout.
    pub fn set_write_timeout(&self, timeout: Option<std::time::Duration>) -> io::Result<()> {
        self.socket.set_write_timeout(timeout)
    }

    /// Check if closed.
    pub fn is_closed(&self) -> bool {
        self.closed.load(Ordering::SeqCst)
    }
}

impl Transport for UdpListener {
    fn send_to(&self, data: &[u8], addr: &dyn Addr) -> Result<()> {
        if self.closed.load(Ordering::SeqCst) {
            return Err(TransportError::Closed);
        }

        // Convert to SocketAddr
        let socket_addr: SocketAddr = if let Some(udp_addr) = addr.as_any().downcast_ref::<UdpAddr>()
        {
            udp_addr.socket_addr()
        } else {
            // Try to parse from string
            addr.addr_string()
                .parse()
                .map_err(|_| TransportError::InvalidAddress)?
        };

        self.socket
            .send_to(data, socket_addr)
            .map_err(TransportError::Io)?;
        Ok(())
    }

    fn recv_from(&self, buf: &mut [u8]) -> Result<(usize, Box<dyn Addr>)> {
        if self.closed.load(Ordering::SeqCst) {
            return Err(TransportError::Closed);
        }

        let (n, addr) = self.socket.recv_from(buf).map_err(TransportError::Io)?;
        Ok((n, Box::new(UdpAddr::new(addr))))
    }

    fn local_addr(&self) -> Box<dyn Addr> {
        Box::new(self.local_addr.clone())
    }

    fn close(&self) -> Result<()> {
        let _lock = self.close_lock.lock().unwrap();
        self.closed.store(true, Ordering::SeqCst);
        // Note: UdpSocket doesn't have a close method in std
        // The socket will be closed when dropped
        Ok(())
    }
}

// Allow cloning the address for injection (testing)
impl UdpListener {
    /// Inject a packet (for testing). Not supported for real UDP.
    pub fn inject_packet(&self, _data: &[u8], _from: &dyn Addr) {
        // No-op for real UDP transport
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_udp_listener_bind() {
        let listener = UdpListener::bind("127.0.0.1:0").unwrap();
        assert!(listener.port() > 0);
        assert!(!listener.is_closed());
    }

    #[test]
    fn test_udp_listener_send_recv() {
        let listener1 = UdpListener::bind("127.0.0.1:0").unwrap();
        let listener2 = UdpListener::bind("127.0.0.1:0").unwrap();

        let addr2 = listener2.local_addr();

        // Send from listener1 to listener2
        let data = b"hello world";
        listener1.send_to(data, addr2.as_ref()).unwrap();

        // Receive on listener2
        let mut buf = [0u8; 1024];
        let (n, from) = listener2.recv_from(&mut buf).unwrap();

        assert_eq!(&buf[..n], data);
        assert_eq!(from.addr_string(), listener1.local_addr().addr_string());
    }

    #[test]
    fn test_udp_listener_multiple_peers() {
        let server = UdpListener::bind("127.0.0.1:0").unwrap();
        let client1 = UdpListener::bind("127.0.0.1:0").unwrap();
        let client2 = UdpListener::bind("127.0.0.1:0").unwrap();

        let server_addr = server.local_addr();

        // Both clients send to server
        client1.send_to(b"from client1", server_addr.as_ref()).unwrap();
        client2.send_to(b"from client2", server_addr.as_ref()).unwrap();

        // Server receives both
        let mut buf = [0u8; 1024];
        let mut received = Vec::new();

        for _ in 0..2 {
            let (n, _from) = server.recv_from(&mut buf).unwrap();
            received.push(String::from_utf8_lossy(&buf[..n]).to_string());
        }

        assert!(received.contains(&"from client1".to_string()));
        assert!(received.contains(&"from client2".to_string()));
    }

    #[test]
    fn test_udp_listener_close() {
        let listener = UdpListener::bind("127.0.0.1:0").unwrap();
        assert!(!listener.is_closed());

        listener.close().unwrap();
        assert!(listener.is_closed());

        // Operations should fail after close
        let result = listener.send_to(b"test", &UdpAddr::parse("127.0.0.1:1234").unwrap());
        assert!(matches!(result, Err(TransportError::Closed)));
    }
}
